drivers/peci/peci_mchp_xec.c

```diff
@@ -0,0 +1,358 @@
+	if (k_sem_take(&data->tx_lock, PECI_IO_DELAY * tx_buf->len)) {
+		return -ETIMEDOUT;
+	}
+#else
+	wait_timeout = tx_buf->len;
+	while (!(base->STATUS1 & MCHP_PECI_STS1_EOF)) {
+		k_busy_wait(PECI_IO_DELAY);
+
+		if (!wait_timeout) {
+			LOG_WRN("Tx timeout\n");
+			base->CONTROL = MCHP_PECI_CTRL_FRST;
+			return -ETIMEDOUT;
+		}
+	}
+#endif
+	return 0;
+}
+
+static int peci_xec_read(struct device *dev, struct peci_msg *msg)
+{
+	int i;
+	int ret;
+	u8_t tx_fcs;
+	u8_t bytes_rcvd;
+	const struct peci_xec_config *config = dev->config->config_info;
+	struct peci_buf *rx_buf = &msg->rx_buffer;
+	PECI_Type *base = config->base;
+
+	/* Attempt to read data from RX FIFO */
+	bytes_rcvd = 0;
+	for (i = 0; i < (rx_buf->len + PECI_FCS_LEN); i++) {
+		/* Check if data available */
+		if (!(base->STATUS2 & MCHP_PECI_STS2_RFE)) {
+			if (i == 0) {
+				/* Get write block FCS just for debug */
+				tx_fcs = base->RD_DATA;
+				LOG_DBG("TX FCS %x", tx_fcs);
+			} else if (i == (rx_buf->len + 1)) {
+				/* Get read block FCS, but don't count it */
+				rx_buf->buf[i-1] = base->RD_DATA;
+			} else {
+				/* Get response */
+				rx_buf->buf[i-1] = base->RD_DATA;
+				bytes_rcvd++;
+			}
+		}
+	}
+
+	/* Check if transaction is as expected */
+	if (rx_buf->len != bytes_rcvd) {
+		LOG_DBG("Incomplete %x vs %x", bytes_rcvd, rx_buf->len);
+	}
+
+	/* Once write-read transaction is complete, ensure bus is idle
+	 * before resetting the internal FIFOs
+	 */
+	ret = check_bus_idle(base);
+	if (ret) {
+
+		return ret;
+	}
+
+	/* Reset internal FIFOs for next transaction */
+	base->CONTROL = MCHP_PECI_CTRL_FRST;
+
+	return 0;
+}
+
+static int peci_xec_transfer(struct device *dev, struct peci_msg *msg)
+{
+	int ret;
+	const struct peci_xec_config *config = dev->config->config_info;
+	PECI_Type *base = config->base;
+
+	ret = peci_xec_write(dev, msg);
+	if (ret) {
+		return ret;
+	}
+
+	/* If a PECI transmission is successful, it may or not involve
+	 * a read operation, check if transaction expects a response
+	 */
+	if (msg->rx_buffer.len) {
+		ret = peci_xec_read(dev, msg);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	/* Cleanup */
+	if (base->STATUS1 & MCHP_PECI_STS1_EOF) {
+		base->STATUS1 |= MCHP_PECI_STS1_EOF;
+	}
+
+	/* Check for error conditions and perform bus recovery if necessary */
+	if (base->ERROR) {
+		if (base->ERROR & MCHP_PECI_ERR_RDOV) {
+			LOG_WRN("Read buffer is not empty\n");
+		}
+
+		if (base->ERROR & MCHP_PECI_ERR_WRUN) {
+			LOG_WRN("Write buffer is not empty\n");
+		}
+
+		if (base->ERROR & MCHP_PECI_ERR_BERR) {
+			LOG_WRN("Write buffer is not empty\n");
+		}
+
+		base->ERROR = base->ERROR;
+		LOG_WRN("Transaction error %x\n", base->ERROR);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_PECI_INTERRUPT_DRIVEN
+static void peci_xec_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct peci_xec_config *config = dev->config->config_info;
+	struct peci_xec_data *data = dev->driver_data;
```

Comment:
You can factor out the peci device and the driver data and config since PECI is a single instance block like Kscan

Reply:
Yes, unlikely will have a future XEC chip with 2 PECI buses.